### DISCUSSION TITLE
How to create an arbitrary js Object in C++?

### DISCUSSION THREAD

**Discussion by sasmaster** [emscripten-core/emscripten#18684] [Q&A]
Hi.
I am trying to create a js style dynamic Object in C++ and set  its properties in runtime.
Here is what I do:

```
   	emscripten::val obj = emscripten::val::object();
 	obj.set("a", 1);
	obj.set("b",2); 
```

Then in browser I have this error:

**Import #116 module="env" function="_emval_new_object" error: function import requires a callable**

What am I doing wrong?I have found this example [here](https://web.dev/emscripten-embedding-js-snippets/).

I also tried created an object like this:

```
emscripten::val JSObject = emscripten::val::global("Object");	 
emscripten::val obj = JSObject.new_();
obj.set("a", 1);
obj.set("b",2); 
```

which results in the same error.

**Comment by sbc100** [comment]
Can you share your full link command?   Are you using emscripten's generated JS file to load the wasm?    Are you linking with `-lembind`?

**Reply by sasmaster** [reply]
Yes I am. Here are the linker flags:

`-sINITIAL_MEMORY=134217728 -sMAX_WEBGL_VERSION=2   -sUSE_GLFW=3  -sFULL_ES3 -sEXIT_RUNTIME -lembind `

But, here is one thing I am not sure about: 
The code above is created inside a member function of a C++ class which is not exposed via embind. Can it be somehow related?

**Reply by sbc100** [reply]
How are you loading your wasm file.   `_emval_new_object` is defined in JS and will end up being part of the JS loading code:  https://github.com/emscripten-core/emscripten/blob/350cb5f0f89645ee3a382c4658cc14acd221a24e/src/embind/emval.js#L140-L144. 

The only way I can think of that it wouldn't be defined when the wasm module is loaded is if you are trying to load the module youself maybe?

**Reply by sasmaster** [reply]
Oh sh*t! What a stupid mistake! I have forgot to copy the accompanying .JS file into web dir. Yes, you're right, that file has been missing. Now that error is gone. But I am still stuck with the problem of instantiating a class with object of callbacks in constructor. Any chance you take a look at [this ](https://github.com/emscripten-core/emscripten/discussions/18672)issue? Thanks a lot!.
